File: src/R/economic.toolkit/R/formulas.R
# SPDX-License-Identifier: PMPL-1.0-or-later
# SPDX-FileCopyrightText: 2024-2026 Hyperpolymath Contributors

#' Calculate Price Elasticity of Demand
#'
#' @param quantity Numeric vector of quantity values
#' @param price Numeric vector of price values
#' @param method Character string: 'point' or 'arc'. Default: 'point'
#'
#' @return Numeric elasticity coefficient
#' @export
#'
#' @examples
#' quantity <- c(100, 90, 80, 70)
#' price <- c(10, 11, 12, 13)
#' elasticity(quantity, price, method = "point")
elasticity <- function(quantity, price, method = "point") {
  if (method == "point") {
    dq <- diff(quantity)
    dp <- diff(price)
    elasticities <- (dq / dp) * (price[-length(price)] / quantity[-length(quantity)])
    return(mean(elasticities))
  } else if (method == "arc") {
    q1 <- quantity[1]
    q2 <- quantity[length(quantity)]
    p1 <- price[1]
    p2 <- price[length(price)]
    return(((q2 - q1) / (q2 + q1)) / ((p2 - p1) / (p2 + p1)))
  } else {
    stop("Unknown method: ", method)
  }
}

#' Calculate GDP Growth Rate
#'
#' @param gdp_values Numeric vector of GDP values over time
#' @param periods Optional integer: number of periods for annualization
#'
#' @return Numeric growth rate(s) as percentage
#' @export
#'
#' @examples
#' gdp <- c(100, 102, 105, 108, 110)
#' gdp_growth(gdp)
#' gdp_growth(gdp, periods = 4)
gdp_growth <- function(gdp_values, periods = NULL) {
  growth_rates <- diff(gdp_values) / gdp_values[-length(gdp_values)] * 100

  if (is.null(periods)) {
    return(growth_rates)
  } else {
    total_growth <- (gdp_values[length(gdp_values)] / gdp_values[1]) ^ (1 / periods) - 1
    return(total_growth * 100)
  }
}

#' Calculate Gini Coefficient
#'
#' @param incomes Numeric vector of income values
#'
#' @return Numeric Gini coefficient (0 = perfect equality, 1 = perfect inequality)
#' @export
#'
#' @examples
#' incomes <- c(20000, 30000, 40000, 50000, 100000, 500000)
#' gini_coefficient(incomes)
gini_coefficient <- function(incomes) {
  sorted_incomes <- sort(incomes)
  n <- length(sorted_incomes)
  cumsum_incomes <- cumsum(sorted_incomes)

  gini <- (2 * sum((n - seq_along(sorted_incomes) + 1) * sorted_incomes)) /
    (n * cumsum_incomes[n]) - (n + 1) / n

  return(gini)
}

#' Calculate Lorenz Curve Coordinates
#'
#' @param incomes Numeric vector of income values
#'
#' @return List with cumulative_population and cumulative_income vectors
#' @export
#'
#' @examples
#' incomes <- c(20000, 30000, 40000, 50000, 100000, 500000)
#' curve <- lorenz_curve(incomes)
#' plot(curve$cumulative_population, curve$cumulative_income, type = "l")
lorenz_curve <- function(incomes) {
  sorted_incomes <- sort(incomes)
  n <- length(sorted_incomes)

  cumulative_income <- cumsum(sorted_incomes) / sum(sorted_incomes)
  cumulative_population <- seq_len(n) / n

  # Add origin point
  cumulative_population <- c(0, cumulative_population)
  cumulative_income <- c(0, cumulative_income)

  return(list(
    cumulative_population = cumulative_population,
    cumulative_income = cumulative_income
  ))
}

#' Calculate Compound Annual Growth Rate (CAGR)
#'
#' @param beginning_value Numeric starting value
#' @param ending_value Numeric ending value
#' @param num_periods Integer number of periods
#'
#' @return Numeric CAGR as percentage
#' @export
#'
#' @examples
#' cagr(100, 150, 5)
cagr <- function(beginning_value, ending_value, num_periods) {
  return(((ending_value / beginning_value) ^ (1 / num_periods) - 1) * 100)
}

#' Calculate Period-over-Period Growth Rates
#'
#' @param values Numeric vector of time series values
#'
#' @return Numeric vector of growth rates as percentages
#' @export
#'
#' @examples
#' values <- c(100, 105, 110, 108, 115)
#' growth_rate(values)
growth_rate <- function(values) {
  return(diff(values) / values[-length(values)] * 100)
}

File: src/python/economic_toolkit/formulas.py
# SPDX-License-Identifier: PMPL-1.0-or-later
# SPDX-FileCopyrightText: 2024-2026 Hyperpolymath Contributors

"""
Economic formula functions for Python API.

Provides economic calculation functions that can work in REST or Julia mode.
"""

from typing import List, Union, Optional
import numpy as np
from economic_toolkit.client import EconomicToolkit


def elasticity(
    quantity: Union[List[float], np.ndarray],
    price: Union[List[float], np.ndarray],
    method: str = "point",
) -> float:
    """
    Calculate price elasticity of demand.

    Args:
        quantity: Quantity values
        price: Price values
        method: 'point' or 'arc' elasticity

    Returns:
        Elasticity coefficient
    """
    if isinstance(quantity, list):
        quantity = np.array(quantity)
    if isinstance(price, list):
        price = np.array(price)

    if method == "point":
        # Point elasticity: (dQ/dP) * (P/Q)
        dq = np.diff(quantity)
        dp = np.diff(price)
        elasticities = (dq / dp) * (price[:-1] / quantity[:-1])
        return float(np.mean(elasticities))

    elif method == "arc":
        # Arc elasticity: ((Q2-Q1)/(Q2+Q1)) / ((P2-P1)/(P2+P1))
        q1, q2 = quantity[0], quantity[-1]
        p1, p2 = price[0], price[-1]
        return ((q2 - q1) / (q2 + q1)) / ((p2 - p1) / (p2 + p1))

    else:
        raise ValueError(f"Unknown method: {method}")


def gdp_growth(
    gdp_values: Union[List[float], np.ndarray],
    periods: Optional[int] = None,
) -> Union[float, np.ndarray]:
    """
    Calculate GDP growth rate(s).

    Args:
        gdp_values: GDP values over time
        periods: Number of periods for annualization (default: None, returns period-over-period)

    Returns:
        Growth rate(s) as percentage
    """
    if isinstance(gdp_values, list):
        gdp_values = np.array(gdp_values)

    # Period-over-period growth
    growth_rates = np.diff(gdp_values) / gdp_values[:-1] * 100

    if periods is None:
        return growth_rates if len(growth_rates) > 1 else float(growth_rates[0])

    # Annualized growth rate
    total_growth = (gdp_values[-1] / gdp_values[0]) ** (1 / periods) - 1
    return float(total_growth * 100)


def gini_coefficient(incomes: Union[List[float], np.ndarray]) -> float:
    """
    Calculate Gini coefficient for income distribution.

    Args:
        incomes: List or array of income values

    Returns:
        Gini coefficient (0 = perfect equality, 1 = perfect inequality)
    """
    if isinstance(incomes, list):
        incomes = np.array(incomes)

    # Sort incomes
    sorted_incomes = np.sort(incomes)
    n = len(sorted_incomes)

    # Calculate Gini coefficient
    cumsum = np.cumsum(sorted_incomes)
    return (2 * np.sum((n - np.arange(n)) * sorted_incomes)) / (n * cumsum[-1]) - (n + 1) / n


def lorenz_curve(incomes: Union[List[float], np.ndarray]) -> tuple:
    """
    Calculate Lorenz curve coordinates.

    Args:
        incomes: List or array of income values

    Returns:
        Tuple of (cumulative_population_share, cumulative_income_share)
    """
    if isinstance(incomes, list):
        incomes = np.array(incomes)

    # Sort incomes
    sorted_incomes = np.sort(incomes)
    n = len(sorted_incomes)

    # Calculate cumulative shares
    cumulative_income = np.cumsum(sorted_incomes) / np.sum(sorted_incomes)
    cumulative_population = np.arange(1, n + 1) / n

    # Add origin point
    cumulative_population = np.concatenate([[0], cumulative_population])
    cumulative_income = np.concatenate([[0], cumulative_income])

    return (cumulative_population.tolist(), cumulative_income.tolist())


def cagr(
    beginning_value: float,
    ending_value: float,
    num_periods: int,
) -> float:
    """
    Calculate Compound Annual Growth Rate (CAGR).

    Args:
        beginning_value: Starting value
        ending_value: Ending value
        num_periods: Number of periods

    Returns:
        CAGR as percentage
    """
    return ((ending_value / beginning_value) ** (1 / num_periods) - 1) * 100


def growth_rate(values: Union[List[float], np.ndarray]) -> Union[float, np.ndarray]:
    """
    Calculate period-over-period growth rates.

    Args:
        values: Time series values

    Returns:
        Growth rates as percentages
    """
    if isinstance(values, list):
        values = np.array(values)

    growth_rates = np.diff(values) / values[:-1] * 100
    return growth_rates if len(growth_rates) > 1 else float(growth_rates[0])
